### PR TITLE
[Merged by Bors] - List.rdrop lemmata

### PR DESCRIPTION
**Pull request by Happyves**
Adds the code from
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/List.20lemma.20for.20mathlib

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Happyves** [comment]
It said "able to merge" when I made the request. I don't understand. Help.

File: Mathlib/Data/List/DropRight.lean [hunk 1]
```diff
@@ -247,4 +247,22 @@ theorem rtakeWhile_idempotent : rtakeWhile p (rtakeWhile p l) = rtakeWhile p l :
   rtakeWhile_eq_self_iff.mpr fun _ => mem_rtakeWhile_imp
 #align list.rtake_while_idempotent List.rtakeWhile_idempotent
 
+
+lemma rdrop_add.{u} {α : Type u} {l : List α} (i j : ℕ) :
```
Comment (jcommelin) [inline Mathlib/Data/List/DropRight.lean:251]:
```suggestion
lemma rdrop_add {l : List α} (i j : ℕ) :
```
The type `α` already is available as a variable. In fact `l` is also available, but as explicit argument. So adding it manually to make it implicit is fine.

File: Mathlib/Data/List/DropRight.lean [hunk 2]
```diff
@@ -247,4 +247,22 @@ theorem rtakeWhile_idempotent : rtakeWhile p (rtakeWhile p l) = rtakeWhile p l :
...
+    l.rdrop (i + j)  = (l.rdrop i).rdrop j  := by
+  simp_rw [List.rdrop_eq_reverse_drop_reverse, reverse_reverse,
+           drop_drop, Nat.add_comm]
+
+@[simp]
+lemma rdrop_append_length.{u} {α : Type u} {l₁ l₂ : List α} :
```
Comment (jcommelin) [inline Mathlib/Data/List/DropRight.lean:257]:
```suggestion
lemma rdrop_append_length {l₁ l₂ : List α} :
```

File: Mathlib/Data/List/DropRight.lean [hunk 3]
```diff
@@ -247,4 +247,22 @@ theorem rtakeWhile_idempotent : rtakeWhile p (rtakeWhile p l) = rtakeWhile p l :
...
+    List.rdrop (l₁ ++ l₂) (List.length l₂) = l₁:= by
+  rw [List.rdrop_eq_reverse_drop_reverse, ← length_reverse l₂,
+      reverse_append, drop_left, reverse_reverse]
+
+@[simp]
+lemma rdrop_append.{u} {α : Type u} {l₁ l₂ : List α} (i : ℕ) :
```
Comment (jcommelin) [inline Mathlib/Data/List/DropRight.lean:263]:
```suggestion
lemma rdrop_append {l₁ l₂ : List α} (i : ℕ) :
```

**Comment by jcommelin** [comment]
Please label your PR with `awaiting-review` when you think it is ready for a review again. If it also gets a :heavy_check_mark: from CI, it will then show up on the reviewer queues.

**Comment by kbuzzard** [comment]
You can also resolve the conversation begun by github actions, because it looks like you have implemented the suggestion there.

File: Mathlib/Data/List/DropRight.lean [hunk 4]
```diff
@@ -247,4 +247,22 @@ theorem rtakeWhile_idempotent : rtakeWhile p (rtakeWhile p l) = rtakeWhile p l :
   rtakeWhile_eq_self_iff.mpr fun _ => mem_rtakeWhile_imp
 #align list.rtake_while_idempotent List.rtakeWhile_idempotent
 
+
+lemma rdrop_add {l : List α} (i j : ℕ) :
+    l.rdrop (i + j) = (l.rdrop i).rdrop j := by
+  simp_rw [List.rdrop_eq_reverse_drop_reverse, reverse_reverse,
+           drop_drop, Nat.add_comm]
```
Comment (urkud) [inline Mathlib/Data/List/DropRight.lean:254]: Is there a reason for LHS/RHS inconsistency between `List.drop_drop` and `List.rdrop_add`?

File: Mathlib/Data/List/DropRight.lean [hunk 5]
```diff
@@ -247,4 +247,22 @@ theorem rtakeWhile_idempotent : rtakeWhile p (rtakeWhile p l) = rtakeWhile p l :
   rtakeWhile_eq_self_iff.mpr fun _ => mem_rtakeWhile_imp
 #align list.rtake_while_idempotent List.rtakeWhile_idempotent
 
+
```
Comment (urkud) [inline Mathlib/Data/List/DropRight.lean:250]:
```suggestion
```

File: Mathlib/Data/List/DropRight.lean [hunk 6]
```diff
@@ -247,4 +247,22 @@ theorem rtakeWhile_idempotent : rtakeWhile p (rtakeWhile p l) = rtakeWhile p l :
...
+           drop_drop, Nat.add_comm]
+
+@[simp]
+lemma rdrop_append_length {l₁ l₂ : List α} :
+    List.rdrop (l₁ ++ l₂) (List.length l₂) = l₁:= by
+  rw [List.rdrop_eq_reverse_drop_reverse, ← length_reverse l₂,
+      reverse_append, drop_left, reverse_reverse]
+
+@[simp]
+lemma rdrop_append {l₁ l₂ : List α} (i : ℕ) :
```
Comment (urkud) [inline Mathlib/Data/List/DropRight.lean:263]: `rdrop_append_length_add`? Also, could you please add lemma `rdrop_append_of_le_length` saying `k ≤ length l₂ → List.rdrop (l₁ ++ l₂) k = l₁ ++ List.rdrop l₂ k`?

File: Mathlib/Data/List/DropRight.lean (hunk 4 above)
Comment (grunweg) [reply Mathlib/Data/List/DropRight.lean]: I didn't see one - flipped.

File: Mathlib/Data/List/DropRight.lean (hunk 6 above)
Comment (grunweg) [reply Mathlib/Data/List/DropRight.lean]: Added; my proof can certainly be golfed further.

File: Mathlib/Data/List/DropRight.lean [hunk 7]
```diff
@@ -241,10 +241,30 @@ theorem mem_rtakeWhile_imp {x : α} (hx : x ∈ rtakeWhile p l) : p x := by
   exact mem_takeWhile_imp hx
 #align list.mem_rtake_while_imp List.mem_rtakeWhile_imp
 
-variable (p) (l)
-
-theorem rtakeWhile_idempotent : rtakeWhile p (rtakeWhile p l) = rtakeWhile p l :=
+theorem rtakeWhile_idempotent (p : α → Bool) (l : List α) :
```
Comment (grunweg) [inline Mathlib/Data/List/DropRight.lean:244]: This is a drive-by clean-up: my understanding is this is worthwhile anyway. It also makes the next lines nicer to write, as I can re-use `l`.

**Comment by grunweg** [comment]
I talked with @Happyves offline; they indicated they wouldn't mind my helping out: I have addressed all comments, as far as I can see.
awaiting-review

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Comment by grunweg** [comment]
Thanks for the review!

**Pull request closed without merging**